test/testdialog.c

```diff
@@ -0,0 +1,148 @@
+            SDLTest_CommonLogUsage(state, argv[0], options);
+            return 1;
+        }
+
+        i += consumed;
+    }
+
+    if (SDL_Init(SDL_INIT_VIDEO) < 0) {
+        SDL_Log("SDL_Init failed (%s)", SDL_GetError());
+        return 1;
+    }
+    if (SDL_CreateWindowAndRenderer(640, 480, 0, &w, &r) < 0) {
+        SDL_Log("Failed to create window and/or renderer: %s\n", SDL_GetError());
+        return 1;
+    }
+
+    while (1) {
+        int quit = 0;
+        SDL_Event e;
+        while (SDL_PollEvent(&e)) {
+            if (e.type == SDL_EVENT_QUIT) {
+                quit = 1;
+                break;
+            } else if (e.type == SDL_EVENT_MOUSE_BUTTON_UP) {
+                const SDL_FPoint p = { e.button.x, e.button.y };
+                /*
+                 * Arguments, in order:
+                 * - A function to call when files are chosen (or dialog is canceled, or error happens)
+                 * - A user-managed void pointer to pass to the function when it will be invoked
+                 * - The window to bind the dialog to, or NULL if none
+                 * - A list of filters for the files, see SDL_DialogFileFilter above (not for SDL_ShowOpenFolderDialog)
+                 * - The path where the dialog should start. May be a folder or a file
+                 * - Nonzero if the user is allowed to choose multiple entries (not for SDL_ShowSaveFileDialog)
+                 */
+                if (SDL_PointInRectFloat(&p, &open_file_rect)) {
+                    SDL_ShowOpenFileDialog(callback, NULL, w, filters, "/home/", 1);
+                } else if (SDL_PointInRectFloat(&p, &open_folder_rect)) {
+                    SDL_ShowOpenFolderDialog(callback, NULL, w, "/home/", 1);
+                } else if (SDL_PointInRectFloat(&p, &save_file_rect)) {
+                    SDL_ShowSaveFileDialog(callback, NULL, w, filters, "/home/");
+                }
+            }
+        }
+        if (quit) {
+            break;
+        }
+        SDL_Delay(100);
+
+        SDL_SetRenderDrawColor(r, 0, 0, 0, SDL_ALPHA_OPAQUE);
+        SDL_RenderClear(r);
+
+        SDL_SetRenderDrawColor(r, 255, 0, 0, SDL_ALPHA_OPAQUE);
+        SDL_RenderFillRect(r, &open_file_rect);
+
+        SDL_SetRenderDrawColor(r, 0, 255, 0, SDL_ALPHA_OPAQUE);
+        SDL_RenderFillRect(r, &save_file_rect);
+
+        SDL_SetRenderDrawColor(r, 0, 0, 255, SDL_ALPHA_OPAQUE);
+        SDL_RenderFillRect(r, &open_folder_rect);
+
+        SDL_SetRenderDrawColor(r, 0, 0, 0, SDL_ALPHA_OPAQUE);
+        SDLTest_DrawString(r, open_file_rect.x+5, open_file_rect.y+open_file_rect.h/2, "Open File...");
+        SDLTest_DrawString(r, save_file_rect.x+5, save_file_rect.y+save_file_rect.h/2, "Save File...");
+        SDLTest_DrawString(r, open_folder_rect.x+5, open_folder_rect.y+open_folder_rect.h/2, "Open Folder...");
+
+        SDL_RenderPresent(r);
+    }
+
+    SDL_DestroyRenderer(r);
+    SDL_DestroyWindow(w);
+    SDL_Quit();
```

Comment:
There was a small leak because of cached text textures.
```suggestion
    SDLTest_CleanupTextDrawing();
    SDL_Quit();
```